Merge `proc_macro_` expansion feature gates as `proc_macro_hygiene`

jebrosen opened the pull request:
Merges `proc_macro_mod`, `proc_macro_expr`, `proc_macro_non_items`, and `proc_macro_gen` into a single feature: `proc_macro_hygiene`. These features are not all blocked on implementing macro hygiene *per se*, but rather on interactions with hygiene that have not been entirely resolved.

rust-highfive commented:
r? @nikomatsakis

(rust_highfive has picked a reviewer for you, use r? to override)

rust-highfive commented:
The job `x86_64-gnu-llvm-3.9` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/401102095) ([raw log](https://api.travis-ci.org/v3/job/401102095/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
travis_time:start:tidy
tidy check
[00:04:13] * 553 error codes
[00:04:13] * highest error code: E0710
[00:04:13] Expected a gate test for the feature 'macros_2'.
[00:04:13] Hint: create a failing test file named 'feature-gate-macros_2.rs'
[00:04:13]       in the 'ui' test suite, with its failures due to
[00:04:13]       missing usage of #![feature(macros_2)].
[00:04:13] Hint: If you already have such a test and don't want to rename it,
[00:04:13]       you can also add a // gate-test-macros_2 line to the test file.
[00:04:13] tidy error: Found 1 features without a gate test.
[00:04:14] some tidy checks failed
[00:04:14] 
[00:04:14] 
[00:04:14] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/tidy" "/checkout/src" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "--no-vendor" "--quiet"
[00:04:14] 
[00:04:14] 
[00:04:14] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test src/tools/tidy
[00:04:14] Build completed unsuccessfully in 0:00:54
[00:04:14] Build completed unsuccessfully in 0:00:54
[00:04:14] Makefile:79: recipe for target 'tidy' failed
[00:04:14] make: *** [tidy] Error 1

The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:08b03d28
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
---
travis_time:end:014e3d1a:start=1530922926684824381,finish=1530922926691254464,duration=6430083
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:0032d840
$ head -30 ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
head: cannot open ‘./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers’ for reading: No such file or directory
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:023af003
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

Mark-Simulacrum commented:
Cc @alexcrichton

alexcrichton commented:
Hm I'd personally prefer to leave these as smaller feature gates in the sense that we started as one large `proc_macro` feature gate but it wasn't realistic to stabilize everything all at once so we scaled it back. Each feature indicates something that will likely require a separate RFC and/or separate discussion to stabilize, and otherwise bucketing all of them gives either the false impression of "when anything is stabilized everything is stabilized" or that there's so much functionality inside it'll never be stable

SergioBenitez commented:
@alexcrichton Continuing down this path makes both developing and using crates with proc-macros (with their full abilities) _very_ painful. For instance, once Rocket migrates to using proc-macros universally, users will need to enable at least four of these features to use Rocket at all. This is, in fact, the impetus for this PR.

Accepting this PR has no implications on whether everything has to be stabilized at once or not. All of the existing features will continue to exist, and you can continue to subset features into different gates. Crates asking users to enable 'macros_2' can stop doing so as soon as the subset of features the crate requires are stable.

I see no downsides to this PR, aside from needing to maintain the feature flag, but I do a see a strong win in terms of usability for both end-users and developers.

alexcrichton commented:
@SergioBenitez I agree from an author's point of view that it's much easier to just write `macro_2`, but it's also much easier to purely write `#![nightly]` and get everything as well! There's a line somewhere at "where should the granularity of opt-in be" and I personally feel like "by unit of stabilization" makes the most sense in terms of signal to noise. It provides signal to us of what to stabilize as well as signal to users as to how many features they're actually opting-in to. It can indeed look like noise if you're opting in to a ton of features, but I see what as somewhat intentional personaly. 

In any case though @nikomatsakis is listed for review here, and mine isn't the only opinion!

jebrosen commented:
My concern with making the granularity of opt-in "by unit of stabilization" is that those units keep changing. For example, last time I tried it locally, #52081 broke `proc_macro2`, apparently only by adding the `proc_macro_diagnostic` feature. If the additional features were documented I wouldn't be bothered so much, but most of the features point to #38356 where there is usually no discussion, tracking, or explanation of them.

alexcrichton commented:
@jebrosen that breakage (and recent breakage) has been from splitting up the `proc_macro` feature, which I don't think has happened a huge amount historically (FWIW)

bors commented:
:umbrella: The latest upstream changes (presumably #52081) made this pull request unmergeable. Please resolve the merge conflicts.

nikomatsakis commented:
r? @alexcrichton

nikomatsakis commented:
> In any case though @nikomatsakis is listed for review here, and mine isn't the only opinion!

Heh, not any more! 😛

mikeyhew commented:
@alexcrichton 

> I agree from an author's point of view that it's much easier to just write macro_2, but it's also much easier to purely write `#![nightly]` and get everything as well! There's a line somewhere at "where should the granularity of opt-in be" and I personally feel like "by unit of stabilization" makes the most sense in terms of signal to noise. It provides signal to us of what to stabilize as well as signal to users as to how many features they're actually opting-in to. It can indeed look like noise if you're opting in to a ton of features, but I see what as somewhat intentional personally.

Writing `#![nightly]` seems like a far cry from what is being proposed here. In this case, all of the features are part of a single user-facing feature — or maybe two features, procedural macros and declarative macros. I guess maybe all of those features mean something to whoever implemented them in the compiler, but for anyone trying to use them, either for Rocket, or just to test them out, they're some complicated incantation that you have to write at the top of your crate.

By the way, can someone explain what all of those features *do*? None of them have explanations in the unstable book, and all six of the proc_macro ones point to the same tracking issue.

jebrosen commented:
I looked more closely at the individual gates and have some more thoughts on this. It ended up being a bit of a brain dump instead of any particular argument, so bear with me.

For a macro crate author: `proc_macro_diagnostic`, `proc_macro_qoute`, and `proc_macro_span` are `proc_macro` library features that a proc_macro crate must enable in order to use them. `decl_macro` must be enabled to define decl macros. I think it's probably fair to require macro crate authors to list these out separately, but it will be frustrating if more changes happen to them.

For someone *using* a proc_macro crate, things are a lot muddier. `use_extern_macros` is relatively understandable and hopefully stabilizing soon. `proc_macro_path_invoc` appears to be a style choice and opt-in for the macro consumer. But the requirements for `proc_macro_mod`, `proc_macro_expr`, `proc_macro_non_items`, and `proc_macro_gen` are more likely imposed on a consumer crate because of something in the macro crate.

According to my understanding of the errors defined in libsyntax/ext/expand.rs and what activates them:

* `proc_macro_mod` - allows *applying* proc_macro attributes to a module.
* `proc_macro_expr` - allows *applying* proc_macro attributes to a statement or expression.
* `proc_macro_non_items` - allows *expanding* proc_macros to non-items (expressions, patterns, types, or statements). This certainly matters for bang-style/inline macros; I can't think off the top of my head what kind of attribute macro would need to do this.
* `proc_macro_gen` - allows *expanding* proc_macros to anything containing a module or a macro definition.

Whether these gates are necessary or not is always (usually?) determined by the macro, but the consumer has to enable the features that support the macros they happen to use. For example, if Rocket's macros were all reimplemented today to perform exactly as the current plugin-based ones, I believe most application crates would need `proc_macro_non_items` and `proc_macro_gen` (in addition to `use_extern_macros` and `decl_macro`).

Side note: if `use_extern_macros` stabilizes, I suspect consumer crates might not need `decl_macro` in order to use a `decl_macro`.

alexcrichton commented:
I think that all makes sense to me @jebrosen! One possibility that may imply to me is that we could have fine-grained gates for *definitions* of procedural macros but not for the invocation of procedural macros. That should help improve the user experience I believe?

jebrosen commented:
I think I have to take back what I said about macros implying the gates they need, at least in general.
For example, consider a macro that replaces every "foo" with "bar" when it appears in any `Ident`. The answer to "which feature gates have to be enabled to use this macro?" is "it depends on the macro input, which is up to the macro consumer".

On the other hand, Rocket's current collection of macros are all targeted at specific use cases:  expanding to a `vec!` call with some name replacement, expanding to other macro calls, attributes that expand to item and/or macro definitions, etc. For these macros, the question "which feature gates have to be enabled to use this macro?" mostly have a clear answer on a case-by-case basis.

From a "wants" perspective, it might be nice for macro authors to be able to opt-in to allow `proc_macro_non_items` on individual macro definition for macros that are intended to be used that way, *and also* for macro users to opt-in to `proc_macro_non_items` in other instances. But that might be too unwieldy to implement to be worth it. It also slightly erodes the sense of "users realize what they're getting into when they enable all these features", but on the other hand downstream crates don't usually have to enable features that upstream crates use anyway.

SergioBenitez commented:
> I think that all makes sense to me @jebrosen! One possibility that may imply to me is that we could have fine-grained gates for definitions of procedural macros but not for the invocation of procedural macros. That should help improve the user experience I believe?

That sounds great to me.

alexcrichton commented:
@jebrosen it's true yeah that a macro's definition doesn't always imply where it's going to be expanded, but as you've found I think it's safe to assume that most macros have a specific context they're expected to expand to, so I think a lot of macros could benefit from the strategy of "the macro is tagged with features it can expand to"

SergioBenitez commented:
@alexcrichton Do I understand correctly that you would be amenable to merging a PR that:

  1. Requires proc-macro crate authors to enable fine-grained features like `proc_macro_non_items`, `proc_macro_gen`, and `decl_macro` in order to use the corresponding functionality.
  2. Does _not_ require consumers of proc-macro crates to enable detailed features. Instead, consumer crates can enable `macros_2` to use any proc-macro or decl-macro.

alexcrichton commented:
@SergioBenitez correct yeah! Something like that I think would reasonably inform users about the dangers of using all these features while hopefully not being too cumbersome as well

jebrosen commented:
I'm trying another formulation of this:

*`macros_2` enables Macros 2.0 features that would otherwise need to be enabled individually on macro consumer crates "because an error message said I had to"*

* `proc_macro_mod`, `proc_macro_gen`, `proc_macro_non_items`, `proc_macro_expr`: *Consumer* gate, but often necessitated by a particular proc_macro. **Included in the `macros_2` feature.**
* `proc_macro_path_invoc`: Consumer gate, but *no proc_macro should necessitate its use*. **Not included.**
* `proc_macro_diagnostic`, `proc_macro_quote`, `proc_macro_span`: *proc_macro crates* can opt-in to these unstable library APIs. **Not included.**
* `decl_macro`: Only the *defining crate* needs the feature; consumer crates are already covered by the stabilization of `use_extern_macros`. **Not included.**

Note that this proposal does not even touch on "tagging" macros with whitelisted expansion types, which I think is a better ideal solution but is also unreasonably complex to implement in the compiler.

Under this proposal, Rocket application crates would need to enable `#![feature(macros_2, decl_macro)]`, which is shorter than both the ~4 individual proc_macro features and over the current compiler plugin incantations. Rocket's current plugin-based macros expand to decl_macro definitions, which is why the consumer crates need `decl_macro` in this case.

@SergioBenitez how do you feel about leaving decl_macro out for these reasons?

jebrosen commented:
As far as I can tell, #53459 doesn't address any of the features imposed on consumer crates by the macro author, which is what this feature request has become. Only `proc_macro_gen` is being partially stabilized; `proc_macro_gen` for macro generation, `proc_macro_mod`, `proc_macro_non_items`, and `proc_macro_expr` are still gated.

jebrosen commented:
Rebased and updated according to the new vision for this feature I outlined above. Hopefully the CI tests pass; I had unrelated failures when I tried locally.

rust-highfive commented:
The job `x86_64-gnu-llvm-5.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/420588357) ([raw log](https://api.travis-ci.org/v3/job/420588357/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[00:52:41] ....................................................................................................
[00:52:44] ....................................................................................................
[00:52:47] ....i...............................................................................................
[00:52:50] ....................................................................................................
[00:52:53] .....................................................iiiiiiiii......................................
[00:52:59] ....................................................................................................
[00:53:03] ....................................................................................................
[00:53:06] ..................................i.................................................................
[00:53:10] ....................................................................................i.i..ii.........
---
travis_time:start:test_run-pass-fulldeps
Check compiletest suite=run-pass-fulldeps mode=run-pass (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
[01:01:36] 
[01:01:36] running 97 tests
[01:03:30] ......................................F.......................test [run-pass] run-pass-fulldeps/myriad-closures.rs has been running for over 60 seconds
[01:04:59] thread 'main' panicked at 'Some tests failed', tools/compiletest/src/main.rs:497:22
[01:04:59] failures:
[01:04:59] 
[01:04:59] ---- [run-pass] run-pass-fulldeps/macros2.rs stdout ----
[01:04:59] 
[01:04:59] 
[01:04:59] error: auxiliary build of "/checkout/src/test/run-pass-fulldeps/auxiliary/macros2.rs" failed to compile: 
[01:04:59] status: exit code: 1
[01:04:59] command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/run-pass-fulldeps/auxiliary/macros2.rs" "--target=x86_64-unknown-linux-gnu" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass-fulldeps/macros2/auxiliary" "-Crpath" "-O" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass-fulldeps/macros2/auxiliary"
[01:04:59] ------------------------------------------
[01:04:59] 
[01:04:59] ------------------------------------------
[01:04:59] stderr:
[01:04:59] stderr:
[01:04:59] ------------------------------------------
[01:04:59] warning: unused imports: `Span`, `quote`
[01:04:59]   --> /checkout/src/test/run-pass-fulldeps/auxiliary/macros2.rs:16:18
[01:04:59]    |
[01:04:59] 16 | use proc_macro::{quote, Span, TokenStream};
[01:04:59]    |                  ^^^^^  ^^^^
[01:04:59]    = note: #[warn(unused_imports)] on by default
[01:04:59] 
[01:04:59] 
[01:04:59] error[E0658]: use of unstable library feature 'proc_macro_quote' (see issue #38356)
[01:04:59]   --> /checkout/src/test/run-pass-fulldeps/auxiliary/macros2.rs:16:18
[01:04:59]    |
[01:04:59] 16 | use proc_macro::{quote, Span, TokenStream};
[01:04:59]    |
[01:04:59]    = help: add #![feature(proc_macro_quote)] to the crate attributes to enable
[01:04:59] 
[01:04:59] error: aborting due to previous error
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

rust-highfive commented:
The job `x86_64-gnu-llvm-5.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/420607492) ([raw log](https://api.travis-ci.org/v3/job/420607492/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[00:47:38] ....................................................................................................
[00:47:41] ....................................................................................................
[00:47:43] ....i...............................................................................................
[00:47:46] ....................................................................................................
[00:47:49] .....................................................iiiiiiiii......................................
[00:47:54] ....................................................................................................
[00:47:58] ....................................................................................................
[00:48:01] ..................................i.................................................................
[00:48:04] ....................................................................................i.i..ii.........
---
travis_time:start:test_run-pass-fulldeps
Check compiletest suite=run-pass-fulldeps mode=run-pass (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
[00:55:49] 
[00:55:49] running 97 tests
[00:57:38] ........................................F..........................test [run-pass] run-pass-fulldeps/myriad-closures.rs has been running for over 60 seconds
[00:59:02] failures:
[00:59:02] 
[00:59:02] ---- [run-pass] run-pass-fulldeps/macros2.rs stdout ----
[00:59:02] 
[00:59:02] 
[00:59:02] error: compilation failed!
[00:59:02] status: exit code: 1
[00:59:02] command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/run-pass-fulldeps/macros2.rs" "--target=x86_64-unknown-linux-gnu" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass-fulldeps/macros2/a" "-Crpath" "-O" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass-fulldeps/macros2/auxiliary"
[00:59:02] ------------------------------------------
[00:59:02] 
[00:59:02] ------------------------------------------
[00:59:02] stderr:
[00:59:02] stderr:
[00:59:02] ------------------------------------------
[00:59:02] error: cannot find macro `d!` in this scope
[00:59:02]    |
[00:59:02]    |
[00:59:02] 50 | d!(df);
[00:59:02]    | ^ help: you could try the macro: `m`
[00:59:02] error: aborting due to previous error
[00:59:02] 
[00:59:02] 
[00:59:02] ------------------------------------------
---
[00:59:02] 
[00:59:02] thread 'main' panicked at 'Some tests failed', tools/compiletest/src/main.rs:497:22
[00:59:02] 
[00:59:02] 
[00:59:02] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/run-pass-fulldeps" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass-fulldeps" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--mode" "run-pass" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-5.0/bin/FileCheck" "--host-rustcflags" "-Crpath -O -Zunstable-options " "--target-rustcflags" "-Crpath -O -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python2.7" "--lldb-python" "/usr/bin/python2.7" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "5.0.0\n" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--llvm-components" "" "--llvm-cxxflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"
[00:59:02] 
[00:59:02] 
[00:59:02] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test
[00:59:02] Build completed unsuccessfully in 0:15:23
[00:59:02] Build completed unsuccessfully in 0:15:23
[00:59:02] make: *** [check] Error 1
[00:59:02] Makefile:58: recipe for target 'check' failed

The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:1cb2216a
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

alexcrichton commented:
This seems like a reasonable idea to me in the sense that there's likely a group of gates we have today that we'll be able to ungate all at once when hygiene is sorted out. To that end it seems reasonable to gate them all with one gate because they're basically all dependent on the same feature.

As I [mentioned above](https://github.com/rust-lang/rust/pull/52121#issuecomment-407448094) `proc_macro_non_items` and `proc_macro_expr` definitely fall in this category. I think `proc_macro_gen` falls in this category too as it's basically the hygiene of macros defining macros. I don't think we want to include `proc_macro_mod` to be included in this list though because it gates expanding entire modules, which isn't so much tied to hygiene as it is "do we even want to support this?"

I think we'd also probably want to basically avoid making a new feature and instead just renaming all these existing feature gates under one name. Something like `#![feature(proc_macro_hygiene2)]` or something like that may be an appropriate name as all these features are basically gated on the same thing.

bors commented:
:umbrella: The latest upstream changes (presumably #53604) made this pull request unmergeable. Please resolve the merge conflicts.

jebrosen commented:
> As I mentioned above `proc_macro_non_items` and `proc_macro_expr` definitely fall in this category. I think `proc_macro_gen` falls in this category too as it's basically the hygiene of macros defining macros.

I think I'm mostly with you on that. `proc_macro_expr` is a bit weird, since it gates applying an attribute to a whole expression -- that seems like more of a syntax decision than hygiene to me.

> I don't think we want to include `proc_macro_mod` to be included in this list though because it gates expanding entire modules, which isn't so much tied to hygiene as it is "do we even want to support this?"

I agree that it's not obviously hygiene-related, but note that both expanding to modules and passing a module into a non-attribute proc-macro are already allowed; only the application of an attribute proc-macro *to* a module is what's blocked by the gate.

> I think we'd also probably want to basically avoid making a new feature and instead just renaming all these existing feature gates under one name. Something like #![feature(proc_macro_hygiene2)] or something like that may be an appropriate name as all these features are basically gated on the same thing.

I think that makes sense, once the appropriate set of features is clear.

SergioBenitez commented:
Hoping to move this forward: it sounds like we've settled on introducing a new feature gate that supplants all existing feature gates related to proc-macro hygiene. @jebrosen, @alexcrichton: Does that sound right? If so, I propose the name `macros_2_hygiene` for this feature.

If we're all in agreement, it looks like @jebrosen should:

  1. Remove the following feature gates:
    * `proc_macro_mod`
    * `proc_macro_expr`
    * `proc_macro_non_items`
    * `proc_macro_gen`
  2. Introduce a `macros_2_hygiene` feature that gates all functionality previously gates by the above features.

petrochenkov commented:
None of these features is blocked on hygiene, except for `proc_macro_gen`.

- `proc_macro_mod` is, IIRC, about expanding inner attributes and expanding non-inline modules (`mod foo;`). I'm not sure why `#[my_macro] mod m { ... }` with outer attribute and inline module is still gated, I'm not aware of any reasons to not stabilize it.
- `proc_macro_expr`/`proc_macro_non_items` have no reasons to be gated as well and can be stabilized after some implementation audit. Proc macros have well-defined "copy-paste" `Span::call_site()` hygiene, this includes "unhygienic" use of local variables.
IIRC, @nrc assumed that additional feature gates would prevent interactions of that copy-paste hygiene with local variables, but that's not true - you can easily expand an item and inject tokens from the outside into its body (e.g. https://github.com/rust-lang/rust/issues/50050), causing the same interactions.

What is really blocked on hygiene is `Span::def_site()` and `decl_macro` (which is also blocked on a ton of other stuff).

SergioBenitez commented:
I don't think what I've proposed runs in conflict with what you're suggesting, @petrochenkov. If that's the position that will be adopted - that copy/paste hygiene _is_ the hygiene of proc-macros, like it wort - then the one feature gate can be removed later.

petrochenkov commented:
I'm not strongly against introducing an umbrella feature, I'm just telling that it's highly unlikely that the stabilization will happen in one step.

>that copy/paste hygiene *is* the hygiene of proc-macros

For already stable macros - sure, in general - not at all.

nrc commented:
My feelings are that we should limit uses of any new macros until we have a really good understanding of the hygiene model. I'm not sure how to achieve that. Using macros in expression position will require more hygiene manipulation (and is more susceptible to hygiene errors) than in item position, and as such I think there is more risk.

nrc commented:
As for one feature gate or many, I don't really care. As long as all the components are tracked in issues and so forth, it should be fine.

jebrosen commented:
> If we're all in agreement, it looks like @jebrosen should:
>
>    Remove the following feature gates:
>    * proc_macro_mod
>    * proc_macro_expr
>    * proc_macro_non_items
>    * proc_macro_gen
>    Introduce a macros_2_hygiene feature that gates all functionality previously gates by the above features.
>
> Edit: proc_macro_hygiene would also be a good name

I've implemented this approach, pending a local test run. If some of these gates can be partially stabilized that's great too, but it sounds like there is not a consensus on that.

rust-highfive commented:
The job `x86_64-gnu-llvm-5.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/435394436) ([raw log](https://api.travis-ci.org/v3/job/435394436/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[00:51:25] .................................................................................................... 2000/4314
[00:51:29] ..................................................................i................................. 2100/4314
[00:51:31] .................................................................................................... 2200/4314
[00:51:34] .................................................................................................... 2300/4314
[00:51:37] ...............iiiiiiiii............................................................................ 2400/4314
[00:51:42] .................................................................................................... 2600/4314
[00:51:46] ...................................................................................................i 2700/4314
[00:51:49] .................................................................................................... 2800/4314
[00:51:52] ...........................................................i.i..ii.................................. 2900/4314
---
travis_time:start:test_codegen
Check compiletest suite=codegen mode=codegen (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
[01:04:50] 
[01:04:50] running 106 tests
[01:04:53] i..ii..iii....i....i........i..iii...........i.....i....ii...i.i.ii..............i...ii..ii.i....iii 100/106
[01:04:53] test result: ok. 76 passed; 0 failed; 30 ignored; 0 measured; 0 filtered out
[01:04:53] 
[01:04:53]  finished in 3.345
[01:04:53] travis_fold:end:test_codegen
---
travis_time:start:test_run-pass-fulldeps
Check compiletest suite=run-pass-fulldeps mode=run-pass (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
[01:05:55] 
[01:05:55] running 97 tests
[01:07:50] ...................................F.F...................F......test [run-pass] run-pass-fulldeps/myriad-closures.rs has been running for over 60 seconds
[01:09:57] .....F.............F.............
[01:09:57] thread 'main' panicked at 'Some tests failed', tools/compiletest/src/main.rs:499:22
[01:09:57] 
[01:09:57] ---- [run-pass] run-pass-fulldeps/macro-quote-cond.rs stdout ----
[01:09:57] 
[01:09:57] 
[01:09:57] error: auxiliary build of "/checkout/src/test/run-pass-fulldeps/auxiliary/cond_plugin.rs" failed to compile: 
[01:09:57] status: exit code: 1
[01:09:57] command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/run-pass-fulldeps/auxiliary/cond_plugin.rs" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Zui-testing" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass-fulldeps/macro-quote-cond/auxiliary" "-Crpath" "-O" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass-fulldeps/macro-quote-cond/auxiliary"
[01:09:57] ------------------------------------------
[01:09:57] 
[01:09:57] ------------------------------------------
[01:09:57] stderr:
[01:09:57] stderr:
[01:09:57] ------------------------------------------
[01:09:57] {"message":"procedural macros cannot be expanded to expressions (see issue #38356)","cnstable feature was used.\n\nErroneous code example:\n\n```compile_fail,E658\n#[repr(u128)] // error: use of unstable library feature 'repr128'\nenum Foo {\n    Bar(u64),\n}\n```\n\nIf you're using a stable or a beta version of rustc, you won't be able to use\nany unstable features. In order to do so, please switch to a nightly version of\nrustc (by using rustup).\n\nIf you're using a nightly version of rustc, just add the corresponding feature\nto be able to use it:\n\n```\n#![feature(repr128)]\n\n#[repr(u128)] // ok!\nenum Foo {\n    Bar(u64),\n}\n```\n"},"level":"error","spans":[{"file_name":"/checkout/src/test/run-pass-fulldeps/auxiliary/cond_plugin.rs","byte_start":1440,"byte_end":1470,"line_start":41,"line_end":41,"column_start":13,"column_end":43,"is_primary":true,"text":[{"text":"            quote!(if $test { $rhs } else)","highlight_start":13,"highlight_end":43}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[{"message":"add #![feature(proc_macro_hygiene)] to the crate attributes to enable","code":null,"level":"help","spans":[],"children":[],"rendered":null}],"rendered":"error[E0658]: procedural macros cannot be expanded to expressions (see issue #38356)\n  --> /checkout/src/test/run-pass-fulldeps/auxiliary/cond_plugin.rs:41:13\n   |\nLL |             quote!(if $test { $rhs } else)\n   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n   |\n   = help: add #![feature(proc_macro_hygiene)] to the crate attributes to enable\n\n"}
[01:09:57] {"message":"aborting due to 2 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to 2 previous errors\n\n"}
[01:09:57] {"message":"For more information about this error, try `rustc --explain E0658`.","code":null,"level":"","spans":[],"children":[],"rendered":"For more information about this error, try `rustc --explain E0658`.\n"}
[01:09:57] ------------------------------------------
[01:09:57] 
[01:09:57] thread '[run-pass] run-pass-fulldeps/macro-quote-cond.rs' panicked at 'explicit panic', tools/compiletest/src/runtest.rs:3267:9
[01:09:57] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[01:09:57] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[01:09:57] 
[01:09:57] ---- [run-pass] run-pass-fulldeps/macro-quote-test.rs stdout ----
[01:09:57] 
[01:09:57] error: auxiliary build of "/checkout/src/test/run-pass-fulldeps/auxiliary/hello_macro.rs" failed to compile: 
[01:09:57] status: exit code: 1
[01:09:57] command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/run-pass-fulldeps/auxiliary/hello_macro.rs" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Zui-testing" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass-fulldeps/macro-quote-test/auxiliary" "-Crpath" "-O" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass-fulldeps/macro-quote-test/auxiliary"
[01:09:57] ------------------------------------------
[01:09:57] 
[01:09:57] ------------------------------------------
[01:09:57] stderr:
[01:09:57] stderr:
[01:09:57] ------------------------------------------
[01:09:57] {"message":"procedural macros cannot be expanded to expressions (see iss/     quote!({\nLL | |         fn hello() {}\nLL | |         macro_rules! m { ($$($$t:tt)*) => { $$($$t)* } }\nLL | |         m!(hello());\nLL | |     })\n   | |______^\n   |\n   = help: add #![feature(proc_macro_hygiene)] to the crate attributes to enable\n\n"}
[01:09:57] {"message":"aborting due to previous error","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to previous error\n\n"}
[01:09:57] {"message":"For more information about this error, try `rustc --explain E0658`.","code":null,"level":"","spans":[],"children":[],"rendered":"For more information about this error, try `rustc --explain E0658`.\n"}
[01:09:57] ------------------------------------------
[01:09:57] 
[01:09:57] thread '[run-pass] run-pass-fulldeps/macro-quote-test.rs' panicked at 'explicit panic', tools/compiletest/src/runtest.rs:3267:9
[01:09:57] 
[01:09:57] 
[01:09:57] ---- [run-pass] run-pass-fulldeps/proc-macro/count_compound_ops.rs stdout ----
[01:09:57] 
[01:09:57] error: auxiliary build of "/checkout/src/test/run-pass-fulldeps/proc-macro/auxiliary/count_compound_ops.rs" failed to compile: 
[01:09:57] status: exit code: 1
[01:09:57] command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/run-pass-fulldeps/proc-macro/auxiliary/count_compound_ops.rs" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Zui-testing" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass-fulldeps/proc-macro/count_compound_ops/auxiliary" "-Crpath" "-O" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass-fulldeps/proc-macro/count_compound_ops/auxiliary"
[01:09:57] ------------------------------------------
[01:09:57] 
[01:09:57] ------------------------------------------
[01:09:57] stderr:
[01:09:57] stderr:
[01:09:57] ------------------------------------------
[01:09:57] {"message":"procedural macros cannot be expanded to expressions (see issue #38356)","code":{"code":"E0658","explanation":"\nAn unstable feature was used.\n\nErroneous code example:\n\n```compile_fail,E658\n#[repr(u128)] // error: use of unstable library feature 'repr128'\nenum Foo {\n    Bar(u64),\n}\n```\n\nIf you're using a stable or a beta version of rustc, you won't be able to use\nany unstable features. In order to do so, please switch to a nightly version of\nrustc (by using rustup).\n\nIf you're using a nightly version of rustc, just add the corresponding feature\nto be able to use it:\n\n```\n#![feature(repr128)]\n\n#[repr(u128)] // ok!\nenum Foo {\n    Bar(u64),\n}\n```\n"},"level":"error","spans":[{"file_name":"/checkout/src/test/run-pass-fulldeps/proc-macro/auxiliary/count_compound_ops.rs","byte_start":762,"byte_end":781,"line_start":22,"line_end":22,"column_start":42,"column_end":61,"is_primary":true,"text":[{"text":"    assert_eq!(count_compound_ops_helper(quote!(++ (&&) 4@a)), 3);","highlight_start":42,"highlight_end":61}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[{"message":"add #![feature(proc_macro_hygiene)] to the crate attributes to enable","code":null,"level":"help","spans":[],"children":[],"rendered":null}],"rendered":"error[E0658]: procedural macros cannot be expanded to expressions (see issue #38356)\n  --> /checkout/src/test/run-pass-fulldeps/proc-macro/auxiliary/count_compound_ops.rs:22:42\n   |\nLL |     assert_eq!(count_compound_ops_helper(quote!(++ (&&) 4@a)), 3);\n   |                                          ^^^^^^^^^^^^^^^^^^^\n   |\n   = help: add #![feature(proc_macro_hygiene)] to the crate attributes to enable\n\n"}
[01:09:57] {"message":"aborting due to previous error","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to previous error\n\n"}
[01:09:57] {"message":"For more information about this error, try `rustc --explain E0658`.","code":null,"level":"","spans":[],"children":[],"rendered":"For more information about this error, try `rustc --explain E0658`.\n"}
[01:09:57] ------------------------------------------
[01:09:57] 
[01:09:57] thread '[run-pass] run-pass-fulldeps/proc-macro/count_compound_ops.rs' panicked at 'explicit panic', tools/compiletest/src/runtest.rs:3267:9
[01:09:57] 
[01:09:57] 
[01:09:57] ---- [run-pass] run-pass-fulldeps/proc-macro/hygiene_example.rs stdout ----
[01:09:57] 
[01:09:57] error: auxiliary build of "/checkout/src/test/run-pass-fulldeps/proc-macro/auxiliary/hygiene_example_codegen.rs" failed to compile: 
[01:09:57] status: exit code: 1
[01:09:57] command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/run-pass-fulldeps/proc-macro/auxiliary/hygiene_example_codegen.rs" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Zui-testing" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ruadd #![feature(proc_macro_hygiene)] to the crate attributes to enable","code":null,"level":"help","spans":[],"children":[],"rendered":null}],"rendered":"error[E0658]: procedural macros cannot be expanded to expressions (see issue #38356)\n  --> /checkout/src/test/run-pass-fulldeps/proc-macro/auxiliary/hygiene_example_codegen.rs:22:5\n   |\nLL |     quote!(hello_helper!($input))\n   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n   |\n   = help: add #![feature(proc_macro_hygiene)] to the crate attributes to enable\n\n"}
[01:09:57] {"message":"procedural macros cannot be expanded to statements (see issue #38356)","code":{"code":"E0658","explanation":"\nAn unstable feature was used.\n\nErroneous code example:\n\n```compile_fail,E658\n#[repr(u128)] // error: use of unstable library feature 'repr128'\nenum Foo {\n    Bar(u64),\n}\n```\n\nIf you're using a stable or a beta version of rustc, you won't be able to use\nany unstable features. In order to do so, please switch to a nightly version of\nrustc (by using rustup).\n\nIf you're using a nightly version of rustc, just add the corresponding feature\nto be able to use it:\n\n```\n#![feature(repr128)]\n\n#[repr(u128)] // ok!\nenum Foo {\n    Bar(u64),\n}\n```\n"},"level":"error","spans":[{"file_name":"/checkout/src/test/run-pass-fulldeps/proc-macro/auxiliary/hygiene_example_codegen.rs","byte_start":964,"byte_end":1278,"line_start":29,"line_end":35,"column_start":5,"column_end":6,"is_primary":true,"text":[{"text":"    quote! {","highlight_start":5,"highlight_end":13},{"text":"        extern crate hygiene_example; // This is never a conflict error","highlight_start":1,"highlight_end":72},{"text":"        let string = format!(\"hello {}\", $input);","highlight_start":1,"highlight_end":50},{"text":"        //^ `format!` always resolves to the prelude macro,","highlight_start":1,"highlight_end":60},{"text":"        //| even if a different `format!` is in scope where `hello!` is used.","highlight_start":1,"highlight_end":78},{"text":"        hygiene_example::print(&string)","highlight_start":1,"highlight_end":40},{"text":"    }","highlight_start":1,"highlight_end":6}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[{"message":"add #![feature(proc_macro_hygiene)] to the crate attributes to enable","code":null,"level":"help","spans":[],"children":[],"rendered":null}],"rendered":"error[E0658]: procedural macros cannot be expanded to statements (see issue #38356)\n  --> /checkout/src/test/run-pass-fulldeps/proc-macro/auxiliary/hygiene_example_codegen.rs:29:5\n   |\nLL | /     quote! {\nLL | |         extern crate hygiene_example; // This is never a conflict error\nLL | |         let string = format!(\"hello {}\", $input);\nLL | |         //^ `format!` always resolves to the prelude macro,\nLL | |         //| even if a different `format!` is in scope where `hello!` is used.\nLL | |         hygiene_example::print(&string)\nLL | |     }\n   | |_____^\n   |\n   = help: add #![feature(proc_macro_hygiene)] to the crate attributes to enable\n\n"}
[01:09:57] {"message":"aborting due to 2 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to 2 previous errors\n\n"}
[01:09:57] {"message":"For more information about this eror: use of unstable library feature 'repr128'\nenum Foo {\n    Bar(u64),\n}\n```\n\nIf you're using a stable or a beta version of rustc, you won't be able to use\nany unstable features. In order to do so, please switch to a nightly version of\nrustc (by using rustup).\n\nIf you're using a nightly version of rustc, just add the corresponding feature\nto be able to use it:\n\n```\n#![feature(repr128)]\n\n#[repr(u128)] // ok!\nenum Foo {\n    Bar(u64),\n}\n```\n"},"level":"error","spans":[{"file_name":"/checkout/src/test/run-pass-fulldeps/auxiliary/proc_macro_def.rs","byte_start":715,"byte_end":750,"line_start":22,"line_end":22,"column_start":5,"column_end":40,"is_primary":true,"text":[{"text":"    quote!(fn $name() -> bool { true })","highlight_start":5,"highlight_end":40}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[{"message":"add #![feature(proc_macro_hygiene)] to the crate attributes to enable","code":null,"level":"help","spans":[],"children":[],"rendered":null}],"rendered":"error[E0658]: procedural macros cannot be expanded to expressions (see issue #38356)\n  --> /checkout/src/test/run-pass-fulldeps/auxiliary/proc_macro_def.rs:22:5\n   |\nLL |     quote!(fn $name() -> bool { true })\n   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n   |\n   = help: add #![feature(proc_macro_hygiene)] to the crate attributes to enable\n\n"}
[01:09:57] {"message":"procedural macros cannot be expanded to expressions (see issue #38356)","code":{"code":"E0658","explanation":"\nAn unstable feature was used.\n\nErroneous code example:\n\n```compile_fail,E658\n#[repr(u128)] // error: use of unstable library feature 'repr128'\nenum Foo {\n    Bar(u64),\n}\n```\n\nIf you're using a stable or a beta version of rustc, you won't be able to use\nany unstable features. In order to do so, please switch to a nightly version of\nrustc (by using rustup).\n\nIf you're using a nightly version of rustc, just add the corresponding feature\nto be able to use it:\n\n```\n#![feature(repr128)]\n\n#[repr(u128)] // ok!\nenum Foo {\n    Bar(u64),\n}\n```\n"},"level":"error","spans":[{"file_name":"/checkout/src/test/run-pass-fulldeps/auxiliary/proc_macro_def.rs","byte_start":859,"byte_end":872,"line_start":27,"line_end":27,"column_start":5,"column_end":18,"is_primary":true,"text":[{"text":"    quote!($item)","highlight_start":5,"highlight_end":18}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[{"message":"add #![feature(proc_macro_hygiene)] to the crate attributes to enable","code":null,"level":"help","spans":[],"children":[],"rendered":null}],"rendered":"error[E0658]: procedural macros cannot be expanded to expressions (see issue #38356)\n  --> /checkout/src/test/run-pass-fulldeps/auxiliary/proc_macro_def.rs:27:5\n   |\nLL |     quote!($item)\n   |     ^^^^^^^^^^^^^\n   |\n   = help: add #![feature(proc_macro_hygiene)] to the crate attributes to enable\n\n"}
[01:09:57] {"message":"procedural macros cannot be expanded to expressions (see issue #38356)","code":{"code":"E0658","explanation":"\nAn unstable feature was used.\n\nErroneous code example:\n\n```compile_fail,E658\n#[repr(u128)] // error: use of unstable library feature 'repr128'\nenum Foo {\n    Bar(u64),\n}\n```\n\nIf you're using a stable or a beta version of rustc, you won't be able to use\nany unstable features. In order to do so, please switch to a nightly version of\nrustc (by using rustup).\n\nIf you're using a nightly version of rustc, just add the corresponding feature\nto be able to use it:\n\n```\n#![feature(repr128)]\n\n#[repr(u128)] // ok!\nenum Foo {\n    Bar(u64),\n}\n```\n"},"level":"error","spans":[{"file_name":"/checkout/src/test/run-pass-fulldeps/auxiliary/proc_macro_def.rs","byte_start":940,"byte_end":952,"line_start":32,"line_end":32,"column_start":5,"column_end":17,"is_primary":true,"text":[{"text":"    quote!(true)","highlight_start":5,"highlight_end":17}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[{"message":"add #![feature(proc_macro_hygiene)] to the crate attributes to enable","code":null,"level":"help","spans":[],"children":[],"rendered":null}],"rendered":"error[E0658]: procedural macros cannot be expanded to expressions (see issue #38356)\n  --> /checkout/src/test/run-pass-fulldeps/auxiliary/proc_macro_def.rs:32:5\n   |\nLL |     quote!(true)\n   |     ^^^^^^^^^^^^\n   |\n   = help: add #![feature(proc_macro_hygiene)] to the crate attributes to enable\n\n"}
[01:09:57] {"message":"procedural macros cannot be expanded to expressions (see issue #38356)","code":{"code":"E0658","explanation":"\nAn unstable feature was used.\n\nErroneous code example:\n\n```compile_fail,E658\n#[repr(u128)] // error: use of unstable library feature 'repr128'\nenum Foo {\n    Bar(u64),\n}\n```\n\nIf you're using a stable or a beta version of rustc, you won't be able to use\nany unstable features. In order to do so, please switch to a nightly version of\nrustc (by using rustup).\n\nIf you're using a nightly version of rustc, just add the corresponding feature\nto be able to use it:\n\n```\n#![feature(repr128)]\n\n#[repr(u128)] // ok!\nenum Foo {\n    Bar(u64),\n}\n```\n"},"level":"error","spans":[{"file_name":"/checkout/src/test/run-pass-fulldeps/auxiliary/proc_macro_def.rs","byte_start":1024,"byte_end":1044,"line_start":37,"line_end":37,"column_start":5,"column_end":25,"is_primary":true,"text":[{"text":"    quote!(return true;)","highlight_start":5,"highlight_end":25}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[{"message":"add #![feature(proc_macro_hygiene)] to the crate attributes to enable","code":null,"level":"help","spans":[],"children":[],"rendered":null}],"rendered":"error[E0658]: procedural macros cannot be expanded to expressions (see issue #38356)\n  --> /checkout/src/test/run-pass-fulldeps/auxiliary/proc_macro_def.rs:37:5\n   |\nLL |     quote!(return true;)\n   |     ^^^^^^^^^^^^^^^^^^^^\n   |\n   = help: add #![feature(proc_macro_hygiene)] to the crate attributes to enable\n\n"}
[01:09:57] {"message":"procedural macros cannot be expanded to expressions (see issue #38356)","code":{"code":"E0658","explanation":"\nAn unstable feature was used.\n\nErroneous code example:\n\n```compile_fail,E658\n#[repr(u128)] // error: use of unstable library feature 'repr128'\nenum Foo {\n    Bar(u64),\n}\n```\n\nIf you're using a stable or a beta version of rustc, you won't be able to use\nany unstable features. In order to do so, please switch to a nightly version of\nrustc (by using rustup).\n\nIf you're using a nightly version of rustc, just add the corresponding feature\nto be able to use it:\n\n```\n#![feature(repr128)]\n\n#[repr(u128)] // ok!\nenum Foo {\n    Bar(u64),\n}\n```\n"},"level":"error","spans":[{"file_name":"/checkout/src/test/run-pass-fulldeps/auxiliary/proc_macro_def.rs","byte_start":1116,"byte_end":1127,"line_start":42,"line_end":42,"column_start":5,"column_end":16,"is_primary":true,"text":[{"text":"    quote!($ts)","highlight_start":5,"highlight_end":16}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[{"message":"add #![feature(proc_macro_hygiene)] to the crate attributes to enable","code":null,"level":"help","spans":[],"children":[],"rendered":null}],"rendered":"error[E0658]: procedural macros cannot be expanded to expressions (see issue #38356)\n  --> /checkout/src/test/run-pass-fulldeps/auxiliary/proc_macro_def.rs:42:5\n   |\nLL |     quote!($ts)\n   |     ^^^^^^^^^^^\n   |\n   = help: add #![feature(proc_macro_hygiene)] to the crate attributes to enable\n\n"}
[01:09:57] {"message":"aborting due to 5 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to 5 previous errors\n\n"}
[01:09:57] {"message":"For more information about this error, try `rustc --explain E0658`.","code":null,"level":"","spans":[],"children":[],"rendered":"For more information about this error, try `rustc --explain E0658`.\n"}
[01:09:57] ------------------------------------------
[01:09:57] 
[01:09:57] thread '[run-pass] run-pass-fulldeps/proc_macro.rs' panicked at 'explicit panic', tools/compiletest/src/runtest.rs:3267:9
[01:09:57] 
---
[01:09:57] test result: FAILED. 92 passed; 5 failed; 0 ignored; 0 measured; 0 filtered out
[01:09:57] 
[01:09:57] 
[01:09:57] 
[01:09:57] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/run-pass-fulldeps" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass-fulldeps" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--mode" "run-pass" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-5.0/bin/FileCheck" "--host-rustcflags" "-Crpath -O -Zunstable-options " "--target-rustcflags" "-Crpath -O -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python2.7" "--lldb-python" "/usr/bin/python2.7" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "5.0.0\n" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--llvm-components" "" "--llvm-cxxflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"
[01:09:57] 
[01:09:57] 
[01:09:57] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test
[01:09:57] Build completed unsuccessfully in 0:22:47
[01:09:57] Build completed unsuccessfully in 0:22:47
[01:09:57] make: *** [check] Error 1
[01:09:57] Makefile:58: recipe for target 'check' failed
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

SergioBenitez commented:
Looks great to me! @alexcrichton, what do you think?

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit cb8b72e316250c687699cf75aa32cc2c059568b4 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton cb8b72e316250c687699cf75aa32cc2c059568b4 -->

bors commented:
:hourglass: Testing commit cb8b72e316250c687699cf75aa32cc2c059568b4 with merge 74503eea294e92cb93055163eb77f1d1c4d40dea...

petrochenkov commented on src/libsyntax/feature_gate.rs:
```diff
@@ -441,10 +441,7 @@ declare_features! (
-    (active, proc_macro_expr, "1.27.0", Some(38356), None),
-    (active, proc_macro_non_items, "1.27.0", Some(38356), None),
-    (active, proc_macro_gen, "1.27.0", Some(38356), None),
+    (active, proc_macro_hygiene, "1.27.0", Some(38356), None),
```
Could you move the features into the "removed" list instead of just removing, and also add "subsumed by `proc_macro_hygiene`" as the removal reason?

petrochenkov replied:
@bors r-

petrochenkov commented:
@bors r-
https://github.com/rust-lang/rust/pull/52121#discussion_r221699989

rust-highfive commented:
The job `x86_64-gnu-llvm-5.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/435953984) ([raw log](https://api.travis-ci.org/v3/job/435953984/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain

```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

SergioBenitez commented:
@petrochenkov Looks like we're good.

bors commented:
:umbrella: The latest upstream changes (presumably #54767) made this pull request unmergeable. Please resolve the merge conflicts.

petrochenkov commented:
@bors r+

bors commented:
:pushpin: Commit d3c902f3113575b134641c14a9734b5075d06b09 has been approved by `petrochenkov`

<!-- @bors r=petrochenkov d3c902f3113575b134641c14a9734b5075d06b09 -->

bors commented:
:hourglass: Testing commit d3c902f3113575b134641c14a9734b5075d06b09 with merge 766e21c7d7f9a0469997c714787d8e244df651b3...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.9229), [status-travis](https://travis-ci.org/rust-lang/rust/builds/437543691?utm_source=github_status&utm_medium=notification)
Approved by: petrochenkov
Pushing 766e21c7d7f9a0469997c714787d8e244df651b3 to master...

rust-highfive commented:
📣 Toolstate changed by rust-lang/rust#52121!

Tested on commit rust-lang/rust@766e21c7d7f9a0469997c714787d8e244df651b3.
Direct link to PR: <https://github.com/rust-lang/rust/pull/52121>

💔 clippy-driver on windows: test-pass → test-fail (cc @Manishearth @llogiq @mcarton @oli-obk, @rust-lang/infra).
💔 clippy-driver on linux: test-pass → test-fail (cc @Manishearth @llogiq @mcarton @oli-obk, @rust-lang/infra).
